packages/blade/src/components/Tabs/utils.ts

```diff
@@ -0,0 +1,85 @@
+import React from 'react';
+import type { TabsProps } from './types';
+import { getComponentId } from '~utils/isValidAllowedChildren';
+import { logger, throwBladeError } from '~utils/logger';
+import { useIsomorphicLayoutEffect } from '~utils/useIsomorphicLayoutEffect';
+
+const iconSizeMap = {
+  medium: 'medium',
+  large: 'large',
+} as const;
+
+const badgeSizeMap = {
+  medium: 'small',
+  large: 'medium',
+} as const;
+
+const counterSizeMap = {
+  medium: 'small',
+  large: 'small',
+} as const;
+
+const propRestrictionMap = {
+  Badge: {
+    medium: {
+      size: badgeSizeMap.medium,
+    },
+    large: {
+      size: badgeSizeMap.large,
+    },
+  },
+  Counter: {
+    medium: {
+      size: counterSizeMap.medium,
+    },
+    large: {
+      size: counterSizeMap.large,
+    },
+  },
+} as const;
+
+type TrailingComponents = keyof typeof propRestrictionMap;
+
+const useTabsItemPropRestriction = (
+  trailing: React.ReactNode,
+  tabItemSize: NonNullable<TabsProps['size']>,
+): React.ReactNode => {
+  const [
+    validatedTrailingComponent,
+    setValidatedTrailingComponent,
+  ] = React.useState<React.ReactElement | null>(null);
+
+  // validate and restrict sub component props in trailing prop
+  useIsomorphicLayoutEffect(() => {
+    if (React.isValidElement(trailing)) {
+      const trailingComponentType = getComponentId(trailing) as TrailingComponents;
+      const restrictedProps = propRestrictionMap[trailingComponentType]?.[tabItemSize];
```

Comment:
why do we call it propRestriction? we only need to check on component restriction right?

Reply:
also, the size shouldn't be restricted to `tabItemSize` right they should be mutually exclusive, they are same today but not sure if they'll always have 1:1 mapping?

Reply:
> why do we call it propRestriction? we only need to check on component restriction right?

We don't just check, we override/enhance the component based on the `propRestrictionMap`

> also, the size shouldn't be restricted to tabItemSize right they should be mutually exclusive, they are same today but not sure if they'll always have 1:1 mapping?

Didn't understand what do you mean by mutually exclusive. 
They are not mutually exclusive check above for mappings: 

```ts
const iconSizeMap = {
  medium: 'medium',
  large: 'large',
} as const;

const badgeSizeMap = {
  medium: 'small',
  large: 'medium',
} as const;

const counterSizeMap = {
  medium: 'small',
  large: 'small',
} as const;
```